### PR TITLE
Add unratified Zaamo and Zalrsc extensions

### PR DESCRIPTION
**Pull request by ved-rivos**

### CODE REVIEW

File: disasm/disasm.cc [hunk 1]
```diff
@@ -834,6 +834,34 @@ void disassembler_t::add_instructions(const isa_parser_t* isa)
...
+    DEFINE_XAMO(amoadd_d)
+    DEFINE_XAMO(amoswap_d)
+    DEFINE_XAMO(amoand_d)
+    DEFINE_XAMO(amoor_d)
+    DEFINE_XAMO(amoxor_d)
+    DEFINE_XAMO(amomin_d)
+    DEFINE_XAMO(amomax_d)
+    DEFINE_XAMO(amominu_d)
+    DEFINE_XAMO(amomaxu_d)
+  }
```
Comment (aswaterman) [inline disasm/disasm.cc:832]: The code duplication seems unfortunate.  Can we regroup this with the previous `if` statement into `if (isa->extension_enabled(EXT_ZAAMO) || isa->extension_enabled('A'))`?

File: disasm/disasm.cc [hunk 2]
```diff
@@ -834,6 +834,34 @@ void disassembler_t::add_instructions(const isa_parser_t* isa)
...
+
+  if (isa->extension_enabled(EXT_ZALRSC)) {
+    DEFINE_XLOAD_BASE(lr_w)
+    DEFINE_XAMO(sc_w)
+    DEFINE_XLOAD_BASE(lr_d)
+    DEFINE_XAMO(sc_d)
+  }
```
Comment (aswaterman) [inline disasm/disasm.cc]: Same here.

**Review by aswaterman — CHANGES_REQUESTED** [review]
For consistency, can we follow the resolution of https://github.com/riscv-software-src/riscv-isa-sim/pull/1559#pullrequestreview-1810207205 here, as well?

File: disasm/disasm.cc (hunk 1 above)
Comment (ved-rivos) [reply disasm/disasm.cc:832]: Updated.

File: disasm/disasm.cc (hunk 2 above)
Comment (ved-rivos) [reply disasm/disasm.cc]: updated.

**Comment by ved-rivos** [comment]
> For consistency, can we follow the resolution of [#1559 (review)](https://github.com/riscv-software-src/riscv-isa-sim/pull/1559#pullrequestreview-1810207205) here, as well?

Updated to match the resolution of PR #1559

File: disasm/isa_parser.cc [hunk 3]
```diff
@@ -359,11 +368,11 @@ isa_parser_t::isa_parser_t(const char* str, const char *priv)
     bad_isa_string(str, "'Zcf/Zcd/Zcb/Zcmp/Zcmt' extensions require 'Zca' extension");
   }
 
-  if (extension_table[EXT_ZACAS] && !extension_table['A']) {
+  if (extension_table[EXT_ZACAS] && !extension_table['A'] && !extension_table[EXT_ZAAMO]) {
     bad_isa_string(str, "'Zacas' extension requires 'A' extension");
```
Comment (aswaterman) [inline disasm/isa_parser.cc]: Can you update the error message to match the new logic?

File: disasm/isa_parser.cc [hunk 4]
```diff
@@ -359,11 +368,11 @@ isa_parser_t::isa_parser_t(const char* str, const char *priv)
...
   }
 
-  if (extension_table[EXT_ZABHA] && !extension_table['A']) {
+  if (extension_table[EXT_ZABHA] && !extension_table['A'] && !extension_table[EXT_ZAAMO]) {
     bad_isa_string(str, "'Zabha' extension requires 'A' extension");
```
Comment (aswaterman) [inline disasm/isa_parser.cc]: Same here.  I think it's ready to merge afterwards.

File: disasm/isa_parser.cc (hunk 3 above)
Comment (ved-rivos) [reply disasm/isa_parser.cc]: updated.

File: disasm/isa_parser.cc (hunk 4 above)
Comment (ved-rivos) [reply disasm/isa_parser.cc]: updated.

**Comment by aswaterman** [comment]
Can you rebase to fix the merge conflict?

**Comment by ved-rivos** [comment]
> Can you rebase to fix the merge conflict?

Fixed the merge conflicts.

**Review by aswaterman — APPROVED** [review]
LGTM now.  Thank you, @ved-rivos!

**Comment by scottj97** [comment]
Please don't do foxtrot merges into your branches -- and especially not when you have to add additional commits to clean up the mess it made. Use rebase instead. This has made an unfortunate mess of the master branch now.

**Comment by aswaterman** [comment]
This one's on me.  My approval was based on the earlier commit history, and I merged on the basis of the CI checks passing without looking at the new commit history.

**Pull request merged**